[CR23] Remove prefix "hgp", use "h" for max pumping & generation

flomnes opened the pull request:
### ⚠️ MERGE THIS AFTER #2073 

Using prefix "hgp" made it too complex for the user to create a valid dataset, and too tricky to validate it: need to loop over all MC years, read prefixes "h" and "hgp" to check that minGen[h] <= maxGen[h] for every hour.

flomnes committed:
[CR23] Remove prefix "hgp", use "h" for max pumping & generation

flomnes commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -112,16 +112,16 @@ static void ConvertDailyTSintoHourlyTS(const Matrix<double>::ColumnType& dailyCo
+ ror(timeseriesNumbers),
+ storage(timeseriesNumbers),
+ mingen(timeseriesNumbers),
+ maxHourlyGenPower(timeseriesNumbers),
+ maxHourlyPumpPower(timeseriesNumbers)
```
The real change is here. The rest is deletion.

flomnes committed:
Add mechanism to check TS consistency

flomnes committed:
Fix mingen reset

flomnes committed:
Fix multiple segfaults

flomnes committed:
Fix test segfault

flomnes committed:
Fix regression (wrong number of TS)

flomnes committed:
Add unit tests

flomnes committed:
Simplify

flomnes committed:
Add unit test for single column time-series

flomnes committed:
fix loop

flomnes committed:
Fix build

flomnes committed:
Remove Equalize* stuff

flomnes committed:
Remove hydroModulable & friends (unused)

flomnes committed:
Avoid using resizeMaxPowerTS and resizeGenerationTS

flomnes committed:
Avoid using maxPowerTScount(), use real width instead

flomnes committed:
Fix segfault, remove resizeMaxPowerTS

flomnes committed:
Fix build for tests

guilpier-code commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -223,20 +171,18 @@ bool DataSeriesHydro::LoadMaxPower(const AreaName& areaID, const AnyString& fold
+void DataSeriesHydro::buildHourlyMaxPowerFromDailyTS(
+  const Matrix<double>::ColumnType& DailyMaxGenPower,
+  const Matrix<double>::ColumnType& DailyMaxPumpPower)
 {
```
About : 
```C++
 maxHourlyGenPower.reset(TScount_, HOURS_PER_YEAR);
 maxHourlyPumpPower.reset(TScount_, HOURS_PER_YEAR);
```
**TScount_** is used here to resize the max power TS, knowing that we are building power TS from daily TS sized to **1**.
So, to make very clear, we should have : 
```C++
 maxHourlyGenPower.reset(1, HOURS_PER_YEAR);
 maxHourlyPumpPower.reset(1, HOURS_PER_YEAR);
```
This is more important than it looks : by the time of this comment, there are ambiguities and questions on **TScount_** (is it a number for ready made TS only ? Or does it take into account generated TS ? Some other comments on this PR also involve **TScount_**).
So the less used, the better.

flomnes replied:
OK, yes.

flomnes replied:
@guilpier-code I removed `TScount_`, will now use `TScount()` instead.

flomnes replied:
Actually, I used 1 as you suggested. Using anything else is a **logic error** and introduces regressions.

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -894,15 +894,10 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
-                hydroSeries->setHydroModulability(area);
         }
 
+        hydroSeries->computeTSCount();
```
`hydroSeries->computeTSCount();` : 
The underlying private data member we compute here is **TScount_**
We compute the underlying private data member **TScount_** here, knowing that it will be possibly recomputed in case some **ror** and **storage** are generated (and so resized).
So we can wonder if it worth it leaving it here like that, alone, or computed once for all, after : 
`TSGenerator::ResizeGeneratedTimeSeries` is called.

flomnes replied:
I think it would be best not to store this value, but instead to compute it every time. Fewer errors.

guilpier-code replied:
Yes but we duplicate code.
Ideally simulation should not know if TS are generated or read.
This TS acquisition should be done while loading input data.
This would avoid have further complications like that.
I know that TS generation will disappear, even if we don't know when.
Anyway, this issue will be handled later, in a way or another, if it is gods will.

guilpier-code replied:
Yes but it's slow and complicated. It could be more simple.
Ideally simulation should not know if TS are generated or read.
This TS acquisition should be done while loading input data.
It would allow :
- to would avoid have further complications like that.
- maybe to make some more simplifications in **timeseries-numbers.cpp**

I know that TS generation will disappear, but we don't know when.
Anyway, this issue will be handled later, in a way or another. 
We should create a ticket to study the benefit we would get by taking this path.

guilpier-code commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -145,37 +98,33 @@ void DataSeriesHydro::copyMaxPowerTS(const DataSeriesHydro& source)
-    maxHourlyPumpPower.reset(nbSeries, HOURS_PER_YEAR);
-
-    maxPowerTScount_ = nbSeries;
+    TScount_ = nbSeries;
```
`TScount_ = nbSeries;` : 
Why do we compute **TScount_** in a one way in **computeTSCount()** and in another way here ?
In **computeTSCount()**, we learn that **TScount_** is about 5 hydro TS. Now it's about 2 of them. I'm confused.

flomnes replied:
`computeTSCount` has been removed.

guilpier-code commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -145,37 +98,33 @@ void DataSeriesHydro::copyMaxPowerTS(const DataSeriesHydro& source)
+        maxHourlyGenPower.numberOfColumns(),
+        maxHourlyPumpPower.numberOfColumns()
+    });
+    TScount_ = *std::max_element(nbColumns.begin(), nbColumns.end());
```
**TScount_** : its a max. Shouldn't we rename into **maxTScount_** ?

flomnes replied:
`TScount_` has been removed. This removal simplifies the code quite a bit : we don't need to recompute `TScount_` every now and then.

guilpier-code commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -133,7 +86,7 @@ void DataSeriesHydro::copyGenerationTS(const DataSeriesHydro& source)
     mingen.timeSeries = source.mingen.timeSeries;
 
-    generationTScount_ = source.generationTScount_;
+    TScount_ = source.TScount_;
```
`TScount_ = source.TScount_;` : 
**TScount_** is useful only in the **solver**, not in the **GUI**.
So we don't have to update it here I guess.

flomnes replied:
Yes, I removed it.

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -894,15 +894,10 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
         }
 
+        hydroSeries->computeTSCount();
+
```
_Not exactly in the scope of this PR_

`hydroSeries->resizeTSinDeratedMode(....);` : 
If we're in **derated** mode, what prevents **ror** and **storage** to be resized in case they are generated ? 
If nothing does, should we not expect unwanted consequences ?

flomnes replied:
Yes, derated mode is supposed to average all time-series and run a single MCY. In my opinion, we should still resize & average all TS, included those that are generated.

guilpier-code commented on src/libs/antares/study/parts/hydro/hydromaxtimeseriesreader.cpp:
```diff
@@ -128,9 +128,10 @@ bool HydroMaxTimeSeriesReader::read(const AnyString& folder, bool usedBySolver)
 {
     bool ret = loadDailyMaxPowersAndEnergies(folder, usedBySolver);
     copyDailyMaxEnergy();
+    hydro_.series->computeTSCount();
```
`hydro_.series->computeTSCount();` : 
Please see **.../area/list.cpp**'s review [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2068#discussion_r1592519627) for a comment on that matter.

flomnes replied:
This method has been removed. We now compute the TS count every time it's necessary.

flomnes committed:
Add logs to checks on number of columns, refactor (#2080)

Add more context to allow the user to correct invalid data.

Extract of logs after this PR
```
[2024-05-03 13:43:44][solver][error] Inconsistent number of columns for link fr / gb (direct-capacity: 3, indirect-capacity: 2)
[2024-05-03 13:43:44][solver][error] Please check that all series have the same number of columns or 1 column, or a combination of 1 column and the same number of columns
```

Also, small refactor to split the implementation of function
`TimeSeriesNumbers::CheckNumberOfColumns` in 2.

Using a `std::map` induces a slight change of behavior : unique keys. TS
with the same key will be erased. This is mostly good, for example if
multiple `AreaScratchpad` are created for a given area (parallel
simulations).

flomnes committed:
PR review

flomnes committed:
Use composition for TimeSeriesNumbers (instead of inheritance) (#2085)

Also, use `timeSeriesNumbers[y]` instead of `timeSeriesNumbers[0][y]`.

flomnes committed:
Apply new formatting

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/check-consistency-ts-numbers

flomnes committed:
Merge branch 'fix/check-consistency-ts-numbers' into fix/cr23-remove-scbuilder-hgp

flomnes committed:
Formatting

flomnes committed:
Fix build

flomnes committed:
Remove DataSeriesHydro::computeTSCount()

flomnes committed:
Remove DataSeriesHydro::computeTSCount()

flomnes committed:
Fix logic for legacy studies

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/cr23-remove-scbuilder-hgp

flomnes committed:
Clean up applyToMatrix.hxx
